sparse/src/KokkosSparse_Utils.hpp

```diff
@@ -2330,6 +2330,163 @@ void validateCrsMatrix(int m, int n, const Rowmap &rowmapIn,
+        }
+        col_start = row_start;
+        ncols     = nrows;
+
+        // Rowmap of i-th row-oriented sub-matrix
+        auto A_row_map_sub = Kokkos::subview(
+            A_row_map_h, Kokkos::make_pair(row_start, row_start + nrows + 1));
+
+        // First round: count i-th non-zeros or size of entries_v[i]
+        size_type n_entries = 0;
+        offset_view1d_type first("first", nrows);  // first position per row
+        offset_view1d_type last("last", nrows);    // last position per row
+
+        for (ordinal_type j = 0; j < nrows; j++) {  // loop through each row
+          size_type k1 = A_row_map_sub(j);
+          size_type k2 = A_row_map_sub(j + 1);
+          size_type k;
+          // Assume column indices are sorted in ascending order
+          // Find the position of the start column in the row
+          for (k = k1; k < k2; k++) {
+            ordinal_type col = A_entries_h(k);
+            if (col >= col_start) {
+              break;
+            }
+          }
+          first(j) = k;
+          // Find the position of the last column in the row
+          for (k = k2 - 1; k >= k1; k--) {
+            ordinal_type col = A_entries_h(k);
+            if (col < col_start + ncols) {
+              break;
+            }
+          }
+          last(j) = k;
+          n_entries += (last(j) - first(j) + 1);
+        }
+
+        // Second round:
+        // - create row_map_v[i]
+        // - copy A_entries to entries_v[i] and update entries_v[i] with local
+        // column indices
+        // - copy A_values to values_v[i]
+        row_map_v[i]   = out_row_map_type("row_map_v", nrows + 1);
+        entries_v[i]   = out_entries_type("entries_v", n_entries);
+        values_v[i]    = out_values_type("values_v", n_entries);
+        row_map_h_v[i] = out_row_map_hostmirror_type("row_map_h_v", nrows + 1);
+        entries_h_v[i] = out_entries_hostmirror_type("entries_h_v", n_entries);
+        values_h_v[i]  = out_values_hostmirror_type("values_h_v", n_entries);
+        size_type first_ = 0;
+        for (ordinal_type j = 0; j < nrows; j++) {  // loop through each row
+          size_type nnz     = last(j) - first(j) + 1;
+          row_map_h_v[i](j) = first_;
+          for (size_type k = 0; k < nnz; k++) {
+            entries_h_v[i](first_ + k) = A_entries_h(first(j) + k) - col_start;
+            values_h_v[i](first_ + k)  = A_values_h(first(j) + k);
+          }
+          first_ += nnz;
+        }
+        row_map_h_v[i](nrows) = n_entries;  // last element
+
+        Kokkos::deep_copy(row_map_v[i], row_map_h_v[i]);
+        Kokkos::deep_copy(entries_v[i], entries_h_v[i]);
+        Kokkos::deep_copy(values_v[i], values_h_v[i]);
+
+        DiagBlk_v[i] = crsMat_t("CrsMatrix", nrows, ncols, n_entries,
+                                values_v[i], row_map_v[i], entries_v[i]);
+
+        row_start += nrows;
+      }  // for (ordinal_type i = 0; i < n_blocks; i++)
+    }    // A_nrows >= 1
+  }      // n_blocks > 1
```

Comment:
Same as above. It will help future maintainers to break this into some smaller functions.

Reply:
@e10harvey Done.